cpp/include/cudf/search.hpp

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/types.hpp>
+#include <cudf/column/column.hpp>
+#include <cudf/table/table.hpp>
+
+#include <vector>
+
+namespace cudf {
+namespace experimental {
+
+/**---------------------------------------------------------------------------*
+ * @brief Find smallest indices in a sorted table where values should be 
+ *  inserted to maintain order
+ * 
+ * For each row v in @p values, find the first index in @p t where
+ *  inserting the row will maintain the sort order of @p t
+ * 
+ * Example:
+ * 
+ *  Single column:
+ *      idx      0   1   2   3   4
+ *   column = { 10, 20, 20, 30, 50 }
+ *   values = { 20 }
+ *   result = {  1 }
+ * 
+ *  Multi Column:
+ *      idx        0    1    2    3    4
+ *   t      = {{  10,  20,  20,  20,  20 },
+ *             { 5.0,  .5,  .5,  .7,  .7 },
+ *             {  90,  77,  78,  61,  61 }}
+ *   values = {{ 20 },
+ *             { .7 },
+ *             { 61 }}
+ *   result =  {  3 }
+ * 
+ * @param t             Table to search
+ * @param values        Find insert locations for these values
+ * @param column_order  Vector of column sort order
+ * @param null_precedence The size of a NULL value in comparison to all other
+ * values
+ * @param mr Device memory resource to use for device memory allocation
+ * @return std::unique_ptr<column> A non-nullable column of INT32 elements
+ * containing the insertion points.
+ *---------------------------------------------------------------------------**/
+std::unique_ptr<column> lower_bound(table_view const& t,
+                                    table_view const& values,
+                                    std::vector<order> const& column_order,
+                                    null_order null_precedence,
```

Comment:
Per https://github.com/rapidsai/cudf/issues/2713 we'll want to be able to control this for each column. This will require https://github.com/rapidsai/cudf/issues/2713 be completed first. 
```suggestion
                                    std::vector<null_order> null_precedences,
```

Reply:
Time frame on #2713?  Do we want to delay this PR until 2713 is addressed, do we want to expand the scope of 2713 to include this change, or do we want to create a new PR for after 2713 to conclude this work?  All of the logic change is in row_lexicographic_comparator, so it should be a fairly straightforward follow-on to update this code.

Reply:
I think it's fine to have an additional PR after #2713 is completed. I just wanted to give you a heads up on it.